### DISCUSSION TITLE
Security flaw? (Local storage sessions)

### DISCUSSION THREAD

**Discussion by Tsimpliarakis** [supabase/supabase#20980] [Questions] [answered]
I made an spa using vue3 and i have the users to be able to sign up and sign in using supabse auth. 
But i noticed that it uses localstorage and i can easily copy the localstorage data and transfer them to another PC and browser and login to the user's acount without needing email or password. 
Also if the user logs out and i keep using his last session data and paste it to the localstorage, even tho supabase auth.sessions table is empty, I have access to his account and can upload data without any problem.
EDIT: In the auth.users table you can see the last sign in. If you enter with a stolen localstorage tho and gain access, the last sign in wont update either, so nobody will even notice it.

I have seen some other discussions about localstorage but I don't understand how is this secure, or how i can improve this. I need to make my own server and use http-only cookies?

**Comment by GaryAustin1** [comment, marked as answer]
There is a short window when a jwt can be used on two machines.  This window is controlled by you in settings for the jwt expire time.  Default is 60 minutes.  It can be set down below 5 minutes or so for higher security (at performance cost to refresh more).

If you signout the refresh token is no longer valid so when that window runs out the jwt will no longer be good.  If you use the refresh token on two devices you will very quickly get errors as the refresh token is used twice and then is invalidated.

**Reply by Tsimpliarakis** [reply]
Oh okay I get that! So still a security flaw for 60 mins that I can mitigate down to 5 or less?
Can i somehow make the session to last as long as you have the tab open? And make users able to only be logged in in one device at a time? Idk something to mitigate the risk, because I'm planning of adding payments (demo ones but still).

**Reply by GaryAustin1** [reply]
60 minutes is longest on token without auto refreshing.
You all have to decide if you want to call it a security flaw. It is the way jwt's work with REST APIs and is not unique to Supabase (think Firebase and other Rest API's).

Someone would have to get to the user's local storage and steal the jwt and refresh token AND use them before either the jwt expires OR if needing to refresh it before the already in use system does a refresh or signout.

You can go to an all server side approach with auth and not rely on local storage (I believe).

**Reply by Tsimpliarakis** [reply]
Hmmm I really dont know, ill set the jwt expiration to 5 mins but everywhere i serched i see that storing sessions to localstorage is a terrible idea and you should use http-only cookies.. Anyway thank you for your answers!

**Reply by GaryAustin1** [reply]
If you are doing banking I would not do it that way.  For a note taking app the risk of someone wanting to steal and being able to inject the code into your source code is minimal, and the damage even more minimal.  Then the time they have to use the token is limited so has to be really a planned effort.  IMO.
Supabase, as I said also has server side methods with PKCE as mentioned in this Auth0 post as a mitigation method.
https://auth0.com/docs/secure/security-guidance/data-security/token-storage#browser-local-storage-scenarios

**Reply by Tsimpliarakis** [reply]
What I'm trying to build let's say that is an airbnb clone for my thesis. I want to add demo payments with paypal and i also handle stuff like user's Passport or id numbers/photos. That's why i was concerned.
I am hosting the site to netlify and i say that netlify supoorts functions that i can set to use http only cookies so i might try to implement that, otherwise I'll stick with the local storage and hope that noone will ever find out lol.
